hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/RandomGeneratePrecompiledContract.java

```diff
@@ -0,0 +1,136 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import com.esaulpaugh.headlong.abi.LongType;
+import com.esaulpaugh.headlong.abi.TypeFactory;
+import com.hedera.services.context.properties.GlobalDynamicProperties;
+import com.hedera.services.stream.RecordsRunningHashLeaf;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.swirlds.common.crypto.Hash;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+import org.hyperledger.besu.evm.precompile.AbstractPrecompiledContract;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Supplier;
+
+import static com.hedera.services.exceptions.ValidationUtils.validateTrue;
+import static com.hedera.services.txns.util.RandomGenerateTransitionLogic.randomNumFromBytes;
+
+/**
+ * System contract to generate random numbers. This will generate 256-bit pseudorandom number when no range is provided.
+ * If a given 32-bit range is provided returns a pseudorandom 32-bit integer X belonging to [0, range).
+ * The pseudorandom number is generated using n-3 record's running hash.
+ */
+@Singleton
+public class RandomGeneratePrecompiledContract extends AbstractPrecompiledContract {
+	private static final String PRECOMPILE_NAME = "RandomGenerate";
+	private static final LongType WORD_DECODER = TypeFactory.create("uint32");
+	//random256BitGenerator(uint256)
+	static final int RANDOM_256_BIT_GENERATOR_SELECTOR = 0x267dc6a3;
+	//randomNumberGeneratorInRange(uint32)
+	static final int RANDOM_NUM_IN_RANGE_GENERATOR_SELECTOR = 0x85b4610c;
+	public static final String RANDOM_GENERATE_PRECOMPILE_ADDRESS = "0x169";
+
+	private final GlobalDynamicProperties dynamicProperties;
+	private final Supplier<RecordsRunningHashLeaf> runningHashLeafSupplier;
+
+	@Inject
+	public RandomGeneratePrecompiledContract(final GasCalculator gasCalculator,
+			final GlobalDynamicProperties dynamicProperties,
+			final Supplier<RecordsRunningHashLeaf> runningHashLeafSupplier) {
+		super(PRECOMPILE_NAME, gasCalculator);
+		this.dynamicProperties = dynamicProperties;
+		this.runningHashLeafSupplier = runningHashLeafSupplier;
+	}
+
+	@Override
+	public long gasRequirement(final Bytes bytes) {
+		return dynamicProperties.randomGenerateGasCost();
+	}
+
+	@Override
+	public Bytes compute(final Bytes input, final MessageFrame frame) {
```

Comment:
Is there a transition logic class for `RandomGenerateTransaction`?

To be consistent with other precompiles we should be farming out all logic to such a transition logic class.  The precompile class then dispatches to that logic class, creates a child record, and stores that child record. 

The hip does specify a new transaction.

Reply:
Got it . Yes there is one [here](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/txns/util/RandomGenerateTransitionLogic.java). It is merged in this [PR](https://github.com/hashgraph/hedera-services/pull/3591)